functions/src/functions/mixpanel.tunnel.ts

```diff
@@ -0,0 +1,64 @@
+import http from "http";
+import * as functions from "firebase-functions";
+
+/**
+ * Encode object into query params
+ *
+ * @param obj
+ */
+function serialize(obj: { [key: string | number]: string | number | boolean }) {
+  const str = [];
+  for (const p in obj)
+    if (Object.prototype.hasOwnProperty.call(obj, p)) {
+      str.push(encodeURIComponent(p) + "=" + encodeURIComponent(obj[p]));
+    }
+  return str.join("&");
+}
+
+/**
+ * Proxy request to mixpanel api
+ *
+ * @param request
+ * @param response
+ */
+export function mixpanelTunnelFunction(request: functions.Request, response: functions.Response): void {
+  const mixpanelHost = "api.mixpanel.com";
+
+  let ip = request.ip;
+  if(request.headers['HTTP_X_FORWARDED_FOR']) {
+    ip = <string>request.headers['HTTP_X_FORWARDED_FOR'];
+  } else if(request.headers['HTTP_X_REAL_IP']) {
+    ip = <string>request.headers['HTTP_X_REAL_IP'];
+  }
+
+  try {
+    const mixpanelRequest = http.request({
+      host: mixpanelHost,
+      path: request.url.replace("/tunnel/mixpanel", ""),
+      method: request.method,
+      headers: ip ? { 'X-REAL-IP': ip } : undefined,
+    }, (mixpanelResponse => {
+      let mixpanelResponseBodyStr = '';
+
+      //another chunk of data has been received, so append it to `mixpanelResponseBodyStr`
+      mixpanelResponse.on('data', (chunk) => {
+        mixpanelResponseBodyStr += chunk;
+      });
+
+      //the whole response has been received
+      mixpanelResponse.on('end', () => {
+        for (const headerName in mixpanelResponse.headers) {
+          const excludedHeaders = ['content-encoding', 'content-length', 'transfer-encoding', 'connection', 'access-control-allow-credentials', 'access-control-allow-origin'];
+          if (excludedHeaders.includes(headerName)) {continue}
+          response.setHeader(headerName, mixpanelResponse.headers[headerName]);
+        }
+        response.status(mixpanelResponse.statusCode).send(mixpanelResponseBodyStr).end();
+      });
+    }));
+    mixpanelRequest.write(typeof request.body === "object" ? serialize(request.body) : request.body);
```

Comment:
But it is all in a try catch block. And it will send an http error if something will break inside the try catch block. And also if mixpanel successfully returns an error it won't crash the function and the same response will be sent back